Optimize a jump-around-branch case

tromey opened the pull request:
In my Emacs JIT I noticed that sometimes libjit would emit a
conditional jump to branch around an unconditional jump, like:

    7ffff7fe5160:	0f 85 05 00 00 00    	jne    0x7ffff7fe516b
    7ffff7fe5166:	e9 0e 00 00 00       	jmpq   0x7ffff7fe5179
    7ffff7fe516b:	41 bf 17 00 00 00    	mov    $0x17,%r15d

While this can be worked around by constructing the IL a bit
differently, I was curious about the difficulty of fixing this in the
libjit optimizer.

This patch notices the specific case where a conditional branch simply
jumps around an unconditional branch.  In this case, the condition is
inverted and the extra jump is eliminated.

tromey committed:
Optimize a jump-around-branch case

In my Emacs JIT I noticed that sometimes libjit would emit a
conditional jump to branch around an unconditional jump, like:

    7ffff7fe5160:	0f 85 05 00 00 00    	jne    0x7ffff7fe516b
    7ffff7fe5166:	e9 0e 00 00 00       	jmpq   0x7ffff7fe5179
    7ffff7fe516b:	41 bf 17 00 00 00    	mov    $0x17,%r15d

While this can be worked around by constructing the IL a bit
differently, I was curious about the difficulty of fixing this in the
libjit optimizer.

This patch notices the specific case where a conditional branch simply
jumps around an unconditional branch.  In this case, the condition is
inverted and the extra jump is eliminated.

ademakov commented on jit/jit-block.c:
```diff
@@ -855,6 +916,32 @@ _jit_block_clean_cfg(jit_function_t func)
+					if l7 != 3 then goto .L0
+					goto .L1
+					.L0:
+
```
It looks like that in this situation there might be a label on the unconditional branch that is used from elsewhere as a branch target:

    if l7 != 3 then goto .L0
    .L2:
    goto .L1
    .L0:

Hence as far as I understand your proposed transformation is not safe. A safe transformation would look like this:

    if l7 == 3 then goto .L1
    goto .L0
    .L2:
    goto .L1
    .L0:

If .L2 is not actually used then it should probably be optimized away on a next step of the CFG clean algorithm and we will get the optimal and correct result. But if this does not happen then we might get a pessimization instead of optimization.

ademakov commented on jit/jit-block.c:
```diff
@@ -855,6 +916,32 @@ _jit_block_clean_cfg(jit_function_t func)
+				detach_edge_dst(block->next->succs[0]);
+				attach_edge_dst(block->next->succs[0], block->next->next);
+				block->next->succs[0]->flags = _JIT_EDGE_FALLTHRU;
+				changed = 1;
```
If I understand correctly what you do here then to maintain consistency additional changes are needed: the unconditional branch opcode should be changed to the noop opcode and the block->next->ends_in_dead flag should be reset. But this is irrelevant in the light of the previous comment.

ademakov commented:
Probably an additional condition that the next block has only one predecessor in the CFG (the fall-through edge) is enough to make this safe. Need to be double check this observation though.

tromey committed:
Add tests/unit subdirectory

This adds a tests/unit subdirectory.  This is just a bit of
infrastructure; future patches will add actual test cases here.

tromey commented:
Thanks, I agree with your analysis.

I found this comment, so it seems like more needs to be done:
```
			/* TODO: "hoist branch" part of the Clean algorithm.
			   It is somewhat complicated as libjit conditional
			   branches differ too much from ILOC conditional
			   branches */
```

I'm not sure I really understand this, though.  If a jump goes directly to another jump, can't the operands simply be compared?

tromey commented:
Also I wonder why the tail end of `_jit_block_clean_cfg` doesn't call `eliminate_unreachable` again before looping.  Maybe it's just not necessary now, because things like `merge_empty` do their own bookkeeping?

tromey committed:
Add unit test for CFG transformation

This adds a unit test for the CFG transformation patch.

tromey commented:
I've updated the patch with your recommendations.  I also changed it to check that the block we plan to remove doesn't have its address taken.  I also added some infrastructure so I could write a unit test for this.
